plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ExpressionConverter.java

```diff
@@ -242,10 +248,59 @@ private static Expression or(List<Expression> expressions)
         if (expressions.isEmpty()) {
             return alwaysFalse();
         }
-        if (expressions.size() == 1) {
-            return getOnlyElement(expressions);
+        return combine(expressions, Expressions::or);
+    }
+
+    private static Expression combine(List<Expression> expressions, BiFunction<Expression, Expression, Expression> combiner)
+    {
+        // Build balanced tree for efficient recursive processing that
```

Comment:
Is there actually a performance difference compared to combining them linearly `(a && (b && (c && (d && e))))`. Seems like that would be easier

I guess it depends on depth and if the evaluator has tail call optimizations

Reply:
if they are evaluated recursively (which i'd assume they are), balancing means you are much less likely to exhaust the stack.

Reply:
However, i should maybe adjust the comment here